compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -348,7 +348,14 @@ trait TypeAssigner {
             }
           }
           else {
-            val argTypes = args.tpes
+            // Make sure arguments don't contain the type `pt` itself.
+            // make a copy of the argument if that's the case.
+            // See pos/i6682a.scala for a test case where this matters.
```

Comment:
I see that this makes -Ycheck pass but I don't understand what this accomplishes conceptually, and if I slightly tweak the testcase I can get it to crash on both master and this branch:
```scala
object O{
  val v = [T] => (y:T) => (x:y.type) => 3
  def m = v(identity(v))
}
```
```scala
*** error while checking tests/pos/i6682a.scala after phase typer ***
exception occurred while compiling tests/pos/i6682a.scala
java.lang.AssertionError: assertion failed:
Found:    [T] => (y: T) => y.type => Int
Required: [T] => (y: Any) => y².type => Int

where:    y  is a reference to a value parameter
          y² is a reference to a value parameter
```
So I suspect there's a deeper issue here?

Reply:
The issue is that normally we need to generate fresh copies for every polymorphic type on every usage of that type. But that would lead to a lot of type copying. Type inference generates a fresh copy in ProtoTypes.constrained only if the same type is already in the constraint set. We need something equivalent for tree checkers and other retypers.

Reply:
The new problem seems to be unrelated to the old problem and the fix, and seems to indicate that self application of polymorphic function types is a minefield which should be studied in depth. If have opened #13143 for that.